client/java/src/main/java/com/linecorp/centraldogma/client/FilteredJsonWatcher.java

```diff
@@ -0,0 +1,89 @@
+/*
+ * Copyright 2019 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package com.linecorp.centraldogma.client;
+
+import java.util.Optional;
+import java.util.concurrent.CompletableFuture;
+import java.util.function.BiConsumer;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.node.MissingNode;
+import com.linecorp.centraldogma.common.Revision;
+
+/**
+ * A watcher for a specific part of a JSON file, that piggy-rides on a watcher of the whole JSON file.
+ * It does not make extra requests, and it will not notify listeners of irrelevant changes.
+ */
+public class FilteredJsonWatcher implements Watcher<JsonNode> {
```

Comment:
How about generalizing this a little bit like this:

```java
public interface Watcher<T> {
    ...
    // Maybe throw UnsupportedOperationException when the current Query is 
    // neither Query.ofJsonPath() nor Query.ofJson()?
    Watcher<T> newChild(String... additionalExpressions);
    Watcher<T> newChild(Iterable<String> additionalExpressions);
    <U> Watcher<U> newChild(Function<? super T, ? extends U> function,
                            String... additionalExpressions);
    <U> Watcher<U> newChild(Function<? super T, ? extends U> function,
                            Iterable<String> additionalExpressions);
}

Watcher<JsonNode> parent = dogma.fileWatcher(... Query.ofJson("/foo.json"));
Watcher<JsonNode> child1 = parent.newChild("$.a");
Watcher<Foo> child2 = parent.newChild(node -> new Foo(node), "$.b");
Watcher<JsonNode> grandChild = child1.newChild("$.c");
```

Reply:
If `additionalExpressions` is empty, `newChild()` could be used for any `Watcher`s, even `Watcher<Revision>` from `CentralDogma.repositoryWatcher()`.

Reply:
I considered it and while the fluent API looks nice, I didn't like the fact that it doesn't make sense to refine the query for non-JSON (QueryType only supports IDENTITY and JSON_PATH only).

However, it makes sense to add query transformers (T -> U) that branch from the parent watcher.

Let me consider some refinement here

Reply:
updated. Thanks for the idea